File: examples/src/WPF/CreateAndDialVpnEntryMvvm/CS/CreateAndDialVpnEntryMvvm/MainWindowViewModel.cs
using System.Collections.Generic;
using DotRas;
using GalaSoft.MvvmLight;
using GalaSoft.MvvmLight.Command;
using System;
using System.Collections.ObjectModel;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Net;
using System.Windows;
using GalaSoft.MvvmLight.Messaging;

namespace CreateAndDialVpnEntryMvvm
{
    public class MainWindowViewModel : ViewModelBase
    {
        private readonly RasDialer _dialer;
        private readonly RasPhoneBook _phonebook;
        private IEnumerable<RasEntry> _connectionsDataSource;
        private bool _isDialButtonEnabled;
        private int _selectedConnectionIndex = -1;
        private bool _isBusy;
        private bool _isDisconnectButtonEnabled;
        private RasHandle _handle;
        private const string ADDRESS = "127.0.0.1";
        private const string USERNAME = "user";
        private const string PASSWORD = "pwd";


        /// <summary>
        /// Initializes a new instance of the MainWindowViewModel class.
        /// </summary>
        public MainWindowViewModel()
        {
            _phonebook = new RasPhoneBook();
            _dialer = new RasDialer();
            _dialer.DialCompleted += dialer_DialCompleted;
            _dialer.StateChanged += dialer_StateChanged;

            CreateEntry = new RelayCommand(OnCreateEntry);
            Dial = new RelayCommand<RasEntry>(OnDial);
            Disconnect = new RelayCommand(OnDisconnect);
            WindowInit = new RelayCommand(OnWindowInit);
        }

        public RelayCommand<RasEntry> Dial { get; }
        public RelayCommand CreateEntry { get; }
        public RelayCommand WindowInit { get; }
        public RelayCommand Disconnect { get; }

        public IEnumerable<RasEntry> ConnectionsDataSource
        {
            get => _connectionsDataSource ?? (_connectionsDataSource = new ObservableCollection<RasEntry>());
            private set
            {
                if (_connectionsDataSource == value) return;
                _connectionsDataSource = value;
                RaisePropertyChanged(nameof(ConnectionsDataSource));
            }
        }

        public int SelectedConnectionIndex
        {
            get => _selectedConnectionIndex;
            set
            {
                if (_selectedConnectionIndex == value) return;
                _selectedConnectionIndex = value;
                RaisePropertyChanged(nameof(SelectedConnectionIndex));
                IsDialButtonEnabled = value > 0;
            }
        }

        public bool IsBusy
        {
            get => _isBusy;
            private set
            {
                if (value == _isBusy) return;
                _isBusy = value;
                RaisePropertyChanged(nameof(IsBusy));
                RaisePropertyChanged(nameof(IsConnectionsComboBoxEnabled));
                RaisePropertyChanged(nameof(IsCreateButtonEnabled));
            }
        }

        public bool IsConnectionsComboBoxEnabled => !IsBusy;
        public bool IsCreateButtonEnabled => !IsBusy;

        public bool IsDisconnectButtonEnabled
        {
            get => _isDisconnectButtonEnabled;
            private set
            {
                if (value == _isDisconnectButtonEnabled) return;
                _isDisconnectButtonEnabled = value;
                RaisePropertyChanged(nameof(IsDisconnectButtonEnabled));
            }
        }

        public bool IsDialButtonEnabled
        {
            get => _isDialButtonEnabled;
            set
            {
                if (_isDialButtonEnabled == value) return;
                _isDialButtonEnabled = value;
                RaisePropertyChanged(nameof(IsDialButtonEnabled));
            }
        }

        public override void Cleanup()
        {
            _dialer?.Dispose();
            _phonebook?.Dispose();

            base.Cleanup();
        }

        private void LoadAllUserEntries()
        {
            var entries = (ObservableCollection<RasEntry>)ConnectionsDataSource;
            entries.Clear();

            entries.Add(new RasEntry("Select one..."));
            foreach (var entry in _phonebook.Entries)
            {
                entries.Add(entry);
            }
            AppendStatusText($"Loaded {_phonebook.Entries.Count} entries.");

            SelectedConnectionIndex = 0;
        }

        private void OnCreateEntry()
        {
            _phonebook.Entries.Clear();

            foreach (VpnProtocol value in Enum.GetValues(typeof(VpnProtocol)))
            {
                var device = GetDevice(value);
                if (device != null)
                {
                    var entry = RasEntry.CreateVpnEntry("Connection " + value, ADDRESS, GetStrategy(value), device);
                    _phonebook.Entries.Add(entry);
                    AppendStatusText($"Created {entry.Name}.");

                }
            }

            LoadAllUserEntries();
        }

        private static RasVpnStrategy GetStrategy(VpnProtocol protocol)
        {
            switch (protocol)
            {
                case VpnProtocol.IKEv2:
                    return RasVpnStrategy.IkeV2Only;
                case VpnProtocol.PPTP:
                    return RasVpnStrategy.PptpOnly;
                case VpnProtocol.L2TP:
                    return RasVpnStrategy.L2tpOnly;
                case VpnProtocol.SSTP:
                    return RasVpnStrategy.SstpOnly;
                default:
                    throw new NotSupportedException();
            }
        }

        /// <summary>
        /// Returns the RasDevice for the given protocol
        /// </summary>
        private static RasDevice GetDevice(VpnProtocol protocol)
        {
            string name;
            switch (protocol)
            {
                case VpnProtocol.IKEv2:
                    name = "(IKEV2)".ToLower();
                    break;
                case VpnProtocol.PPTP:
                    name = "(PPTP)".ToLower();
                    break;
                case VpnProtocol.L2TP:
                    name = "(L2TP)".ToLower();
                    break;
                case VpnProtocol.SSTP:
                    name = "(SSTP)".ToLower();
                    break;
                default:
                    throw new NotSupportedException();
            }
            return RasDevice.GetDevices().FirstOrDefault(d => d.DeviceType == RasDeviceType.Vpn && d.Name.ToLower().Contains(name));
        }

        [SuppressMessage("ReSharper", "InconsistentNaming")]
        private enum VpnProtocol
        {
            IKEv2,
            PPTP,
            L2TP,
            SSTP,
        }

        private static void AppendStatusText(string value)
        {
            Messenger.Default.Send(new DialerStateChangedMessage(value));
        }

        private void OnDial(RasEntry entry)
        {
            if (entry == null || entry.Id == Guid.Empty)
            {
                return;
            }

            try
            {
                Messenger.Default.Send(new DialerStateChangedMessage());

                _dialer.EntryName = entry.Name;
                _dialer.PhoneBookPath = entry.Owner.Path;

                _dialer.Credentials = new NetworkCredential(USERNAME, PASSWORD);
                _dialer.SynchronizingObject = new DispatcherSynchronizingObject(Application.Current.Dispatcher);

                _dialer.DialAsync();

                // Make sure the user cannot change the selected connection, and change the buttons that are enabled.
                IsBusy = true;
                IsDialButtonEnabled = false;
                IsDisconnectButtonEnabled = true;
            }
            catch (Exception ex)
            {
                AppendStatusText(ex.ToString());
            }
        }

        private void OnDisconnect()
        {
            if (!_dialer.IsBusy)
            {
                AppendStatusText("Disconnecting...");
                var conn = RasConnection.GetActiveConnections().FirstOrDefault(_ => _.Handle == _handle);
                if (conn != null)
                {
                    conn.HangUp();
                    AppendStatusText("Disconnected!");
                }
                else
                {
                    AppendStatusText("Connection handle not found.");
                }
                IsDisconnectButtonEnabled = false;
                IsBusy = false;
                IsDialButtonEnabled = true;
                return;
            }

            _dialer.DialAsyncCancel();
        }

        private void OnWindowInit()
        {
            _phonebook.SynchronizingObject = new DispatcherSynchronizingObject(Application.Current.Dispatcher);
            _phonebook.Open("phonebook.pbk");

            LoadAllUserEntries();
        }

        private static void dialer_StateChanged(object sender, StateChangedEventArgs e)
        {
            AppendStatusText("State changed: " + e.State);
            if (e.ErrorCode > 0)
            {
                AppendStatusText($"Error code: {e.ErrorCode}. Message: {e.ErrorMessage}");
            }
        }

        private void dialer_DialCompleted(object sender, DialCompletedEventArgs e)
        {
            if (e.Cancelled)
            {
                AppendStatusText("Cancelled!");
            }
            else if (e.Connected)
            {
                _handle = e.Handle;
                AppendStatusText("Connected!");
            }
            else if (e.TimedOut)
            {
                AppendStatusText("Connection attempt timed out!");
            }
            else if (e.Error != null)
            { 
                AppendStatusText(e.Error.ToString());
            }

            IsBusy = e.Connected;
            IsDialButtonEnabled = !e.Connected;
            IsDisconnectButtonEnabled = e.Connected;
        }
    }
}

File: examples/src/WPF/CreateAndDialVpnEntryMvvm/VB/CreateAndDialVpnEntryMvvm/MainWindowViewModel.vb
Imports GalaSoft.MvvmLight
Imports GalaSoft.MvvmLight.Messaging
Imports DotRas
Imports GalaSoft.MvvmLight.Command
Imports System.Collections.ObjectModel
Imports System.Net

Public Class MainWindowViewModel : Inherits ViewModelBase
    Public Const ConnectionName As String = "VPN Connection"

    Private dialer As RasDialer
    Private phoneBook As RasPhoneBook
    Private _connectionsDataSource As IEnumerable(Of RasEntry)
    Private _isCreateButtonEnabled As Boolean = True
    Private _isDisconnectButtonEnabled As Boolean
    Private _isConnectionsComboBoxEnabled As Boolean = True
    Private _isDialButtonEnabled As Boolean
    Private _selectedConnectionIndex As Integer = -1

    Public Sub New()
        Me.phoneBook = New RasPhoneBook()
        AddHandler Me.phoneBook.Changed, AddressOf Me.phoneBook_Changed
        Me.phoneBook.EnableFileWatcher = True

        Me.dialer = New RasDialer()
        AddHandler Me.dialer.StateChanged, AddressOf Me.dialer_StateChanged
        AddHandler Me.dialer.DialCompleted, AddressOf Me.dialer_DialCompleted

        Me.CreateEntry = New RelayCommand(AddressOf Me.OnCreateEntry)
        Me.Dial = New RelayCommand(Of RasEntry)(AddressOf Me.OnDial)
        Me.Disconnect = New RelayCommand(AddressOf Me.OnDisconnect)
        Me.WindowInit = New RelayCommand(AddressOf Me.OnWindowInit)
    End Sub

    Public Property Dial As RelayCommand(Of RasEntry)
    Public Property CreateEntry As RelayCommand
    Public Property WindowInit As RelayCommand
    Public Property Disconnect As RelayCommand

    Public Property ConnectionsDataSource As IEnumerable(Of RasEntry)
        Get
            If Me._connectionsDataSource Is Nothing Then
                Me._connectionsDataSource = New ObservableCollection(Of RasEntry)
            End If

            Return Me._connectionsDataSource
        End Get
        Set(ByVal value As IEnumerable(Of RasEntry))
            If value IsNot Me._connectionsDataSource Then
                Me._connectionsDataSource = value
                Me.RaisePropertyChanged("ConnectionsDataSource")
            End If
        End Set
    End Property

    Public Property SelectedConnectionIndex As Integer
        Get
            Return Me._selectedConnectionIndex
        End Get
        Set(ByVal value As Integer)
            If Me._selectedConnectionIndex <> value Then
                Me._selectedConnectionIndex = value
                Me.RaisePropertyChanged("SelectedConnectionIndex")

                Me.IsDialButtonEnabled = value > 0
            End If
        End Set
    End Property

    Public Property IsConnectionsComboBoxEnabled As Boolean
        Get
            Return Me._isConnectionsComboBoxEnabled
        End Get
        Set(ByVal value As Boolean)
            If (Me._isConnectionsComboBoxEnabled <> value) Then
                Me._isConnectionsComboBoxEnabled = value
                Me.RaisePropertyChanged("IsConnectionsComboBoxEnabled")
            End If
        End Set
    End Property

    Public Property IsCreateButtonEnabled As Boolean
        Get
            Return Me._isCreateButtonEnabled
        End Get
        Set(ByVal value As Boolean)
            If Me._isCreateButtonEnabled <> value Then
                Me._isCreateButtonEnabled = value
                Me.RaisePropertyChanged("IsCreateButtonEnabled")
            End If
        End Set
    End Property

    Public Property IsDisconnectButtonEnabled As Boolean
        Get
            Return Me._isDisconnectButtonEnabled
        End Get
        Set(ByVal value As Boolean)
            If Me._isDisconnectButtonEnabled <> value Then
                Me._isDisconnectButtonEnabled = value
                Me.RaisePropertyChanged("IsDisconnectButtonEnabled")
            End If
        End Set
    End Property

    Public Property IsDialButtonEnabled() As Boolean
        Get
            Return Me._isDialButtonEnabled
        End Get
        Set(ByVal value As Boolean)
            If Me._isDialButtonEnabled <> value Then
                Me._isDialButtonEnabled = value
                Me.RaisePropertyChanged("IsDialButtonEnabled")
            End If
        End Set
    End Property

    Public Overrides Sub Cleanup()
        If dialer IsNot Nothing Then
            dialer.Dispose()
        End If

        If phoneBook IsNot Nothing Then
            phoneBook.Dispose()
        End If

        MyBase.Cleanup()
    End Sub

    Private Sub LoadAllUserEntries()
        Dim entries As ObservableCollection(Of RasEntry) = CType(Me.ConnectionsDataSource, ObservableCollection(Of RasEntry))
        entries.Clear()

        entries.Add(New RasEntry("Select one..."))

        For Each entry As RasEntry In Me.phoneBook.Entries
            entries.Add(entry)
        Next

        Me.IsCreateButtonEnabled = Not Me.phoneBook.Entries.Contains(ConnectionName)
        Me.SelectedConnectionIndex = 0
    End Sub

    Private Sub OnCreateEntry()
        If Me.phoneBook.Entries.Contains(ConnectionName) Then
            Me.IsCreateButtonEnabled = False
        Else
            Dim entry As RasEntry = RasEntry.CreateVpnEntry(ConnectionName, IPAddress.Loopback.ToString(), RasVpnStrategy.Default, RasDevice.GetDeviceByName("(PPTP)", RasDeviceType.Vpn))

            Me.phoneBook.Entries.Add(entry)
        End If
    End Sub

    Private Sub AppendStatusText(ByVal value As String)
        Messenger.Default.Send(Of DialerStateChangedMessage)(New DialerStateChangedMessage(value))
    End Sub

    Private Sub OnDial(ByVal entry As RasEntry)
        If entry IsNot Nothing AndAlso entry.Id = Guid.Empty Then
            Return
        End If

        Try
            Messenger.Default.Send(Of DialerStateChangedMessage)(New DialerStateChangedMessage())

            Me.dialer.EntryName = entry.Name
            Me.dialer.PhoneBookPath = entry.Owner.Path

            Me.dialer.Credentials = New NetworkCredential("Test", "User")
            Me.dialer.SynchronizingObject = New DispatcherSynchronizingObject(Application.Current.Dispatcher)

            Me.dialer.DialAsync()

            ' Make sure the user cannot change the selected connection, and change the buttons that are enabled.
            Me.IsConnectionsComboBoxEnabled = False
            Me.IsDialButtonEnabled = False
            Me.IsDisconnectButtonEnabled = True
        Catch ex As Exception
            Me.AppendStatusText(ex.ToString())
        End Try
    End Sub

    Private Sub OnDisconnect()
        If Not Me.dialer.IsBusy Then
            Return
        End If

        Me.dialer.DialAsyncCancel()
    End Sub

    Private Sub OnWindowInit()
        Me.phoneBook.SynchronizingObject = New DispatcherSynchronizingObject(Application.Current.Dispatcher)
        Me.phoneBook.Open()

        If Me.phoneBook.Entries.Contains(ConnectionName) Then
            ' The entry already exists within the phonebook, disable the button.
            Me.IsCreateButtonEnabled = False
        End If

        Me.LoadAllUserEntries()
    End Sub

    Private Sub phoneBook_Changed(ByVal sender As Object, ByVal e As EventArgs)
        Me.LoadAllUserEntries()
    End Sub

    Private Sub dialer_StateChanged(ByVal sender As Object, ByVal e As StateChangedEventArgs)
        Me.AppendStatusText(e.State.ToString() + Chr(13) + Chr(10))
    End Sub

    Private Sub dialer_DialCompleted(ByVal sender As Object, ByVal e As DialCompletedEventArgs)
        If e.Cancelled Then
            Me.AppendStatusText("Cancelled!")
        ElseIf e.Connected Then
            Me.AppendStatusText("Connected!")
        ElseIf e.TimedOut Then
            Me.AppendStatusText("Connection attempt timed out!")
        ElseIf e.Error IsNot Nothing Then
            Me.AppendStatusText(e.Error.ToString())
        End If

        Me.IsConnectionsComboBoxEnabled = True
        Me.IsDisconnectButtonEnabled = False
        Me.IsDialButtonEnabled = True
    End Sub
End Class